### PR TITLE
Modular file structure

### PR DESCRIPTION
**Pull request by mangkoran**
Hello I am back :D I couldn't stop myself from tinkering with Typst.

This PR is a bigger one. This PR would like to achieve "modularity" (I don't really sure this is the right terms). The main goal is to allow user to keep the template "updated" or at least "in line" with the upstream by adding this repo as a git submodule and use/import the "base" or "style" file (`cv.typ`). This way, user would be able to add their own style rules or even override existing rules in their "content" file (`content.typ`, user can use their own "content" file).

```
user-cv.typ/
├── cv.typ/ // added as git submodule so update can be done easily
|   └── cv.typ
└── content.typ // #import "cv.typ/cv.typ": *
```

To achieve this, several works have been done:

- Update page style rules to be a function. This way, the page style rules function can be redefined to override the rules
- Update content blocks to be a function. This way, the content block functions can be redefined to override the implementation (such as style, content, etc.)

Additionally, there is one WIP stuff which is how we can pass variables for `showAddress` and `showNumber` and read the value in `cv.typ`. Currently I am reading Typst's [dictionary reference](https://typst.app/docs/reference/types/dictionary/) and I think dictionary reassignment could be used here.

As always, please let me know if you have any comments or want me to add something. I am also fine if this PR "contradicts" your ideals for the structure of this template, and I might continue this PR as a fork 😉

### CODE REVIEW

**Comment by mangkoran** [comment]
Commit [`7ce2f81` (#2)](https://github.com/jskherman/cv.typ/pull/2/commits/7ce2f817965f0707c07b9ffded017000a41079c9) now fix variable `showAddress` and `showNumber` passing 🎉🎉🥳🥳

Please let me know if anything should be changed 😉

**Comment by mangkoran** [comment]
Thank you for sparing your time! Let me know before you're going to merge it, as I might have something to add (I am digging into the Typst rabbit hole :D)

**Comment by jskherman** [comment]
Ok 👍, take care to not to fall too deep in the rabbit hole `:3` (just looking at the [plotting libraries](https://github.com/johannes-wolf/typst-canvas/) will make you feel you're staring into the abyss).

**Comment by mangkoran** [comment]
> Ok 👍, take care to not to fall too deep in the rabbit hole :3 (just looking at the [plotting libraries](https://github.com/johannes-wolf/typst-canvas/) will make you feel you're staring into the abyss).

I will try my best to not to. My attention span is so low I easily distracted with interesting things 😅

> Edit: regarding the git submodule, perhaps [making a package](https://github.com/typst/packages) with Typst might be preferable?

Wow this is great. I will try to read it first. But for now, I am going to settle with the submodule approach. If the Typst package is better and doable, we could revisit it later on :D

**Comment by mangkoran** [comment]
Commit [`d053d3f` (#2)](https://github.com/jskherman/cv.typ/pull/2/commits/d053d3f4aa4c6ddc3cecd97006feacd20b60749c) now move the CV data loading out to user-specified "content" file (`content.typ` for example purpose).

**Comment by mangkoran** [comment]
Additionally, I will create a commit to update the readme to reflect this change.

**Comment by jskherman** [comment]
I believe something has changed the line spacings to create loose unordered lists, but I can't seem to pinpoint what it is the code. This results in the projects section being truncated and the have more whitespace between the links section and education section.

image

**Comment by mangkoran** [comment]
> This results in the projects section being truncated and the have more whitespace between the links section and education section.

I couldn't reproduce the wide spacings as you provided. To make sure, please checkout my latest commit. I just reworded the commits so it shouldn't affect anything.

image

**Comment by jskherman** [comment]
Huh, weird. I see, maybe it's on my end. At one point, the output seems to have been the same and then different the next.

Anyways, I believe everything is in order. I would like to suggest though some minor style changes in using curly brackets `{}` instead of `[]` for the surrounding functions for the different "modules".

**Comment by mangkoran** [comment]
> Huh, weird. I see, maybe it's on my end. At one point, the output seems to have been the same and then different the next.

Try to use `typst compile` instead of watch. I also experience "wrong content" compilation. It could be because I am using nvim (see https://github.com/typst/typst/issues/1693).

> I would like to suggest though some minor style changes in using curly brackets {} instead of [] for the surrounding functions for the different "modules", if there are no content at the same level as the inner functions/commands.

I think I could include it in this PR. Could you please left comments in the code so that I could locate the exact styling that you meant?

image

File: cv.typ [hunk 1]
```diff
@@ -1,298 +1,348 @@
...
+                - *Honors*: #edu.honors.join(", ")
+                - *Courses*: #edu.courses.join(", ")
+                // Highlights or Description
+                #for hi in edu.highlights [
+                    - #eval("[" + hi + "]")
+                ]
+            ]
         ]
     ]
 ]
```
Comment (jskherman) [inline cv.typ:150]: Sorry, I meant for example at `cveducation(info)` the part `cveducation(info) = [...]` could be `cveducation(info) = {...}` instead and for similar functions.

Comment (mangkoran) [reply cv.typ]: I have pushed a commit to address this. Could you please take a look on it?

**Comment by mangkoran** [comment]
Let's wait until tomorrow before merging. I still haven't updated the documentation 🤣

**Review by jskherman — COMMENTED** [review]
All good, documentation up next.

**Comment by jskherman** [comment]
> Hi @jskherman. Sorry but today was a busy day for me. I will try to update the docs by tomorrow.

Of course, take your time and whenever you're free.

**Comment by mangkoran** [comment]
Hi @jskherman. I have pushed commits for docs update and files rename. Please take a look on it and let me know if you have any suggestions on it.

File: README.md [hunk 2]
```diff
@@ -20,22 +21,40 @@ See [**sample CV**](cv.pdf) and [@jskherman's CV](https://go.jskherman.com/cv):
...
+apply document style & show CV components and load CV data from a YAML file
+([`example.yml`](example.yml) as example). Inside the content file you can modify several style
+variables and even override existing function implementations to your heart's content.
 
-1. Upload the [`cv.typ`](cv.typ), [`utils.typ`](utils.typ), and [`cv.typ.yml`](cv.typ.yml) files to your Typst project.
-2. In your project, Start editing `cv.typ` for the layout and `cv.typ.yml` for the data.
+### With [Typst CLI](https://github.com/typst/typst) (Recommended)
 
-### With [Typst CLI](https://github.com/typst/typst)
+The recommended usage with Typst CLI is by adding `cv.typ` as [git submodule](https://git-scm.com/book/en/v2/Git-Tools-Submodules). This way upstream changes can be pulled easily.
```
Comment (jskherman) [inline README.md:32]: Perhaps, we could use clearer language here? Someone might confuse `cv.typ` (this GitHub repository) and `cv.typ` (the Typst file).

```suggestion
The recommended usage with Typst CLI is by adding this `cv.typ` repository as a [git submodule](https://git-scm.com/book/en/v2/Git-Tools-Submodules). This way, upstream changes can be pulled easily.
```

**Review by jskherman — CHANGES_REQUESTED** [review]
Just one minor tweak on language before merging. Overall, thank you for the wonderful pull request @mangkoran! Much appreciated.

File: README.md (hunk 2 above)
Comment (mangkoran) [reply README.md]: Updated.

**Review by jskherman — APPROVED** [review]
Merged ✨

**Comment by mangkoran** [comment]
Thank you very much for merging this (and previous) PRs. I am trying to contribute to communities as much as I can do as a payback for such awesome tools they've created. I hope you won't get bored with my future PRs as well 😁

**Comment by jskherman** [comment]
Thank you too, I appreciate the help as well. It's always nice to have some opportunity to interact in the spirit of open-source.

**Pull request merged**